service/history/workflow/cache/cache.go

```diff
@@ -261,9 +261,24 @@ func (c *CacheImpl) makeReleaseFunc(
 				if err != nil || forceClearContext {
 					// TODO see issue #668, there are certain type or errors which can bypass the clear
 					context.Clear()
+					context.Unlock(lockPriority)
+					c.Release(key)
+				} else {
+					isDirty := context.IsDirty()
+					if isDirty {
+						context.Clear()
+						c.logger.Error("Cache encountered dirty mutable state transaction",
+							tag.WorkflowNamespaceID(context.GetWorkflowKey().NamespaceID),
+							tag.WorkflowID(context.GetWorkflowKey().WorkflowID),
+							tag.WorkflowRunID(context.GetWorkflowKey().RunID),
+						)
+					}
+					context.Unlock(lockPriority)
+					c.Release(key)
+					if isDirty {
+						panic("Cache encountered dirty mutable state transaction")
```

Comment:
This will terminate the process. What is the point of doing all the rest (besides logging, of course)?

Reply:
Is it because we need to always release the lock, and the panic may be captured? But then maybe we should not panic here.

Reply:
1. we cannot return error here (passing to API)
2. panic here will make the panic capture from outside turning panic into error and return to API caller